### PR TITLE
fix: add 640th KiB to the available memory

### LINKED ISSUE #3044: Memory map excludes 640th kB
# Describe the bug

The memory map provided by Firecracker is

```
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x000000003fffffff] usable
```

In other words, the first 639 kB of address space exists, and then addresses from 1 MB upwards (in this case up to 1 GB) exist.

The MP table is placed at address 0x9fc00, i.e.. in the last kB of the "normal" 640 kB of base memory -- but according to the memory map, this doesn't exist on the system!

This is harmless for Linux, because Linux's MP table parsing has a bug -- it assumes that there's always 640 kB of base memory.  But it breaks FreeBSD, since FreeBSD looks at the memory map when following the Intel MP spec of "look in the *last kB of base memory*".

## To Reproduce

1. Start Firecracker
2. Look at the memory map printed by the guest kernel

## Expected behaviour

The memory map presented by Firecracker should go up to 640 kB, not 639 kB.

## Environment

Latest Firecracker version checked out from main.  Running on Ubuntu 22.04; guest kernel and rootfs from the "getting started" instructions.

### PR DESCRIPTION
**Pull request by ShadowCurse**
## Changes
Fixes #3044

As described in the issue FC was only announcing 639 KiB as the first memory region available to the guest.
But at the same time FC is placing MP tables in the 640th KiB of the memory.

This PR fixes it by adding an additional entry from `EBDA_START` to `EBDA_START` + 1 KiB to represent mp tables.

Old memory map:
```
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x000000003fffffff] usable
```

New memory map:
```
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x000000000009fc00-0x000000000009ffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x0000000007ffffff] usable
```

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [ ] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: src/vmm/src/arch/x86_64/mod.rs [hunk 1]
```diff
@@ -134,7 +136,7 @@ pub fn configure_system(
         params.hdr.ramdisk_size = initrd_config.size as u32;
     }
 
-    add_e820_entry(&mut params, 0, EBDA_START, E820_RAM)?;
+    add_e820_entry(&mut params, 0, EBDA_END, E820_RAM)?;
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply src/vmm/src/arch/x86_64/mod.rs:139]: I did a test where I added another entry like this:
```rust 
add_e820_entry(&mut params, 0, EBDA_START, E820_RAM)?;
add_e820_entry(&mut params, EBDA_START, EBDA_END, E820_RAM)?;
```
And the result was same as adding 1 entry. (in kernel logs it was reporting as 1 range from 0 to EBDA_END)
I don't think there is a need for 2 entries when 1 covers the correct range anyway.

File: src/vmm/src/arch/x86_64/mod.rs [hunk 2]
```diff
@@ -41,8 +42,11 @@ pub enum ConfigurationError {
     InitrdAddress,
 }
 
-// Where BIOS/VGA magic would live on a real PC.
+// EBDA is located in the last 1 KiB from 0x9FC00 - 0x9FFFF
```
Comment (bchalios) [inline src/vmm/src/arch/x86_64/mod.rs:45]:
```suggestion
// EBDA is located in the last 1 KiB of the first 640KiB of memory, i.e in the range: [0x9FC00, 0x9FFFF]
```

Comment (ShadowCurse) [reply src/vmm/src/arch/x86_64/mod.rs]: added

**Comment by kalyazin** [comment]
In the map comparison (PR description), why are the last entries different (`0x000000003fffffff` vs `0x0000000007ffffff`)? Is this expected?

**Review by bchalios — APPROVED** [review]

**Comment by bchalios** [comment]
> In the map comparison (PR description), why are the last entries different (`0x000000003fffffff` vs `0x0000000007ffffff`)? Is this expected?

Oh, good catch. I totally missed that. I guess the numbers were taken from microVMs with different memory. We should amend that.

**Comment by ShadowCurse** [comment]
Yes, I was using uVM with default 128MB of RAM, while issue was with uVM with 1GB.

**Comment by ShadowCurse** [comment]
> Not sure I get it. I think the issue should exist as well in a microVM with 128MB of memory. The memory map difference should be similar. Am I missing something?

Yes, the comment about different RAM sizes was about Nikita's comment regarding last entries being different.

**Pull request merged**